### PR TITLE
don't make corrupted copies of singleton objects during AST deserialization

### PR DESCRIPTION
**Pull request by vtjnash**
this corruption will now be detected when finalizing the sysimg with an assertion

### CODE REVIEW

File: src/dump.c [hunk 1]
```diff
@@ -2346,14 +2349,21 @@ static void jl_init_restored_modules(jl_array_t *init_order)
...
+    for (i = 0; i < l; i++) {
         jl_value_t *ti = jl_svecref(cache, i);
-        if (ti == NULL) break;
+        if (ti == NULL)
+            break;
         if (ptrhash_get(&backref_table, ti) != HT_NOTFOUND || jl_get_llvm_gv(ti) != 0)
             jl_svecset(cache, ins++, ti);
+        else if (jl_is_datatype(ti)) {
+            jl_value_t *singleton = ((jl_datatype_t*)ti)->instance;
+            if (singleton && (ptrhash_get(&backref_table, singleton) != HT_NOTFOUND || jl_get_llvm_gv(singleton) != 0))
```
Comment (JeffBezanson) [inline src/dump.c]: How can the instance be referenced in the backref table without its type being there?

File: src/dump.c [hunk 2]
```diff
@@ -1853,18 +1855,19 @@ static jl_value_t *jl_deserialize_value_globalref(jl_serializer_state *s)
 
 static jl_value_t *jl_deserialize_value_singleton(jl_serializer_state *s, jl_value_t **loc)
 {
-    int usetable = (s->mode != MODE_AST);
+    if (s->mode == MODE_AST) {
+        jl_datatype_t *dt = (jl_datatype_t*)jl_deserialize_value(s, NULL);
```
Comment (JeffBezanson) [inline src/dump.c]: It seems to me this case would only have happened before for singleton objects that weren't using `LiteralVal_tag`. Are there examples of that?

File: src/dump.c (hunk 1 above)
Comment (vtjnash) [inline src/dump.c]: probably only the `jl_get_llvm_gv` is needed. I just didn't feel like trying to verify / assert that

File: src/dump.c (hunk 2 above)
Comment (vtjnash) [inline src/dump.c]: Yes, there were lots of them (tuples) in the sysimg

Comment (JeffBezanson) [inline src/dump.c]: Ah right, good ol' `Tuple{Void,Void}`.

**Review by JeffBezanson — APPROVED** [review]

**Comment by tkelman** [comment]
This has no test, and `jl_prune_type_cache` does not exist on release-0.5. If you're positive this is a problem on release-0.5 an critical to solve for 0.5.2, please resolve the conflict and open a backport PR against the tk/backports-0.5.2 branch (with `cherry-pick -x -e` and add a cross-reference to this PR in the edited commit message)

**Comment by vtjnash** [comment]
It's tested by julia-debug. If `jl_prune_type_cache` doesn't exist, you can drop the backport of the fix to that function. I don't have a branch of release-0.5 right now, but yes this is a critical issue.

**Comment by tkelman** [comment]
> It's tested by julia-debug.

How so? What is the evidence of whatever problem this fixes?

**Comment by vtjnash** [comment]
That it doesn't detect memory corruption and abort

**Comment by tkelman** [comment]
And what kind of code can trigger this? Have you gotten it to occur on actual code? Even if intermittent, this could really use a test case.

**Comment by vtjnash** [comment]
The sysimg build triggers it and will get caught by the assertion https://github.com/JuliaLang/julia/pull/21206/files#diff-669d4cc5c9c8f4573c5f8d57f5dcab20R1061 when the serializer does a memory scan.

**Comment by tkelman** [comment]
The sysimg builds using julia-debug during `make install`, doesn't it? Why would that not have triggered this?

**Comment by vtjnash** [comment]
Because this PR adds that test?

**Comment by StefanKarpinski** [comment]
In summary, this commit adds an assertion which would raise an error if the bug that the commit fixes were to occur. That seems sufficiently tested. Thanks for the explanation, @vtjnash.

**Pull request merged**